### ISSUE TITLE
mongoose not being traced without await

### ISSUE THREAD

**Issue by radziszp** [googleapis/cloud-trace-nodejs#956]
Hello,
i'm trying to auto trace my mongoose (v 4.13.17, mongo v 3.4.1, node v 8.11) queries. 
Just for testing purpose i've created simple route
```
router.get('/mongo', (req, res) => {
  new Model({ body: 'body'}).save();
  res.send('Request to mongo is completed.');
});
```
in this scenario i receive from DEBUG ` DEBUG TraceApi#createChildSpan: [mongodb-core@2.1.18] Creating phantom child span [mongo-insert] because root span [/app/mongo] was already closed.`
And only HTTP request is traced in [https://console.cloud.google.com/traces](https://console.cloud.google.com/traces)
in `@google-cloud/trace-agent DEBUG TraceWriter#flushBuffer: Flushing traces` array i see only HTTP request.

When i mark this route callback as async and run `await newModel({ body: 'body'}).save();` on  [https://console.cloud.google.com/traces](https://console.cloud.google.com/traces) page, under HTTP request i see `mongo-insert` section.

How can i fix that i can see mongo queries without await?

**Comment by kjin** [comment]
Mongoose appears to only make the query after two `nextTick`s, which would place the start of the query after the request handler has ended completely (since we close the request handler span upon `res.send`). We currently don't allow nesting a span in another span that has already ended, largely because in most cases it's a symptom of async context loss (in addition -- until #913 lands, the span wouldn't be published anyway).

There are a few options here:
1. As a _workaround_, you can capture the query details in a child span by simply delaying the call to `res.send`. If you don't want to `await` the `save()` call, you can wrap `res.send(...)` with a slight delay via `setImmediate`:
    ```js
    router.get('/mongo', (req, res) => {
      new Model({ body: 'body'}).save();
      // two process.nextTicks will also work, at least in mongoose 4.13.17
      setImmediate(() => res.send('Request to mongo is completed.'));
    });
    ```
    This is sort of a hack because it relies `mongoose` internals, and the span duration won't be correct (right now, because we haven't landed #913, the child span when will be truncated when its parent is ended). But it should work if you just want the query details.
2. I can try to move #913 forward, which allows child spans to be ended later than their parents. This doesn't affect start, so I would need to separately add a config option where users can specify that they _do_ want to allow spans to start after their parent span is ended (we won't make this the default behavior). But in using this option, it's possible that some unwanted child spans will spuriously appear.
3. We can consider writing a Mongoose plugin specifically to deal with this. But I can't make any guarantees about how long this will take.

Let me know if the workaround doesn't work in the short term, and we can discuss longer term solutions in more depth. Thanks!

**Comment by radziszp** [comment]
Thank you for such a quick response!

Hmm, if so i'll refactor my code to use `async/await`.
Also I will follow the progress on #913 .
Good luck!

**Comment by ofrobots** [comment]
@kjin now that #913 has landed, what else needs to be done here?

**Comment by kjin** [comment]
We would need to add a config option to allow child spans to be started after a root span has ended.

@radziszp has the above worked for you so far (in at least publishing child spans)?

**Comment by radziszp** [comment]
@kjin i've just installed vesion `3.5.2` but it seems like `js` files from `trace-agent/build` don't have changes from #913

**Comment by kjin** [comment]
@radziszp Sorry for the late response -- we [published 3.6.0](https://github.com/googleapis/cloud-trace-nodejs/releases/tag/v3.6.0) which should include the changes.

**Comment by kjin** [comment]
Closing this for now -- please leave a comment if there are outstanding issues to address.

**Comment by radziszp** [comment]
Hello,
something is still not right :/ 
maybe i'm doing something wrong? Could you please verify?
I've uploaded simple project to github.
After visiting `/mongo-async` everything in tracing mode looks ok, but non-async mongo call aren't shown.
[https://github.com/radziszp/test-tracing](https://github.com/radziszp/test-tracing)


![async](https://user-images.githubusercontent.com/37267385/57967370-426f7c00-795e-11e9-933b-ce364d10e74a.png)
![sync](https://user-images.githubusercontent.com/37267385/57967371-456a6c80-795e-11e9-9fcb-6a0a3eb967c2.png)

**Comment by kjin** [comment]
Sure, let me take a look

**Comment by kjin** [comment]
Even though we added support for ending child spans later than their parent, we still don't support child spans _starting_ after a root span is closed, which is why you're not observing child spans being created in the `/mongo` endpoint. The suggestion I wrote [here](https://github.com/googleapis/cloud-trace-nodejs/issues/956#issuecomment-465271747) has not been implemented yet.

It's possible that I could implement this option, but keep in mind that:
* in normal usage, an arbitrary child span being started after its parent has ended is usually a red flag of a context issue in the application, and we would no longer be able to differentiate the valid and invalid cases where a child span starts after its parent, and
* we usually delete entries from our internal table of parent spans when the span has ended (for memory management purposes), and we would need to introduce some approximate heuristic to extend this deletion instead if we want to create children of spans after they've ended. __(In other words it might not work 100% of the time)__

That being said, is there a particular reason you would want to invoke `find` or other query methods without waiting for the result? And if there is, it seems like it would likely not affect your request latency.
